Add line-circle and circle-circle closest points functions; improve cylinder-cylinder collisions

rburing opened the pull request:
This PR implements Vranek's [*Fast and Accurate Circle-Circle and Circle-Line 3D Distance Computation*](https://www.tandfonline.com/doi/abs/10.1080/10867651.2002.10487552) with the following API:

![line-circle_and_circle-circle_api](https://user-images.githubusercontent.com/229837/213831777-b495cc33-2f73-4ad3-b7ba-659ef1e1316a.png)

The line-circle implementation is based on the one in [`webots`](https://github.com/cyberbotics/webots/blob/21d6eefd7a2f59fa3277788b552587e388dfe0a4/src/ode/ode/src/cylinder_revolution_data.cpp) and the circle-circle implementation is based on the one in [`MinuteTorus`](https://github.com/SonSang/MinuteTorus/blob/1a11cacc8cea6f62d16ffc9919ba35fe998742c9/Circle/CircleDistance.cpp), though both have been significantly modified. The original article is sketchy regarding the line-circle case; I've added details in comments in the code, and answering my [question about the choice of bracketing triplet](https://math.stackexchange.com/questions/4618258/minimum-distance-between-a-line-and-a-circle-in-3d-bracketing-triplet) could improve the implementation.

* Works well even in `precision=single` builds, unlike the alternative implementation of these functions in https://github.com/godotengine/godot/pull/68958.

* Fixes cylinder sliding on cylinder in https://github.com/fabriceci/Godot-Physics-Tests

* Shows significant improvement in the "Face Z" and "Tumbling" cases of https://github.com/Malcolmnixon/PhysicsCollisionTest. I'm not sure why "Face Y" does not work better. I could use some help debugging this.

* Demo to test the circle-circle and line-circle closest points functions directly: [circle-distance-demo.zip](https://github.com/godotengine/godot/files/10471228/circle-distance-demo.zip)

* Demo to test cylinder-cylinder collisions: [CharacterAndCylinderIssue-modified.zip](https://github.com/godotengine/godot/files/10471227/CharacterAndCylinderIssue-modified.zip)

To improve cylinder collisions still further (and implement cylinder-cylinder distance), disk-circle and segment-circle closest points functions should be implemented.

rburing committed:
Add modification of Brent's numerical minimization method using the derivative

rburing committed:
Add separating axis tests for cylinder-cylinder collisions

lawnjelly commented:
Just a note on the allocations - I usually hold my nose at the amount of dynamic allocations Godot does, but here you are using `push_back()` / `append()` to add points to a return Vector, for a function that seems designed to be called realtime.

If possible it would be nice to reduce this to a single allocation.
Some ideas :
* use reserve to pre-reserve the vector with a reasonable number of points
* if you know the maximum possible points, store in a local array or use alloca (on stack), then finally reserve an output vector of the correct size and copy

On the other hand if it is nearly always returning one value, this may not be so important.

rburing commented:
@lawnjelly Thanks, I'll fix that (it's only in the binding, not in the core code that is used by the physics).

Currently the code always returns only one pair (unlike https://github.com/godotengine/godot/pull/68958 which returned more pairs in special cases), which seems to be standard for this particular algorithm. I've indicated in some comments in the code how this could be improved if needed.

rburing commented:
@Riteo Yes, definitely. I meant to post this as a draft; indeed copyright notices / credits still need to be added.

peastman commented:
It looks like some of this code came from Numerical Recipes.  They don't allow using it in open source software.  http://numerical.recipes/licenses/redistribute.html

rburing commented:
@peastman Yes, the minimization and bracketing code in `minimization.cpp`, I've noticed it too.

I'm looking for a solution.

hpvb commented:
@rburing I have an implementation on a branch that does not use the code from Numerical Recipes here: https://github.com/hpvb/godot/commit/56a68d2a6da9425881e837d637bcb82e86d81f2a

Feel free to use it, if you do please add an authored-by line for me or something to the PR :)

rburing commented:
@hpvb Thanks a lot for your help! With your latest commit I still run into the issue of getting caught (stopped) when sliding along the cap of a cylinder (which this PR is supposed to fix), see the demo attached above.

There is also a test like that in https://github.com/fabriceci/Godot-Physics-Tests, but it's already passing in `master`, so I should add a test that fails like the demo.

Could you give the minimization code another look? If needed we could generate some unit tests for the minimization. Once it's ready I'd be happy to rebase my commits on top of yours (and add more detailed license info to `COPYRIGHT.txt`).

rburing commented:
@hpvb My branch [`test_line_circle_distance_minimization`](https://github.com/rburing/godot/tree/test_line_circle_distance_minimization) allows testing the minimization code on the function $$f(t) = A_0 + t\cdot (A_1 + t) + A_2 \cdot \sqrt{A_3 + t\cdot (A_4 + t\cdot A_5)}$$ which (for some $A_i$ ) is the squared distance between a circle and a point $P(t)$ on a line, for $t$ in some interval $[a,b]$.

To test alternative minimization code (like yours), check out my branch, replace `minimization.cpp`, and run the test project [line_circle_distance_minimization_test.zip](https://github.com/godotengine/godot/files/11004881/line_circle_distance_minimization_test.zip). It compares calculated minima with the pre-computed minima found using this PR (stored in `minimization_data.json`), and prints the data where the error in (squared) distance is largest.

Running the test project with your minimization code, I get

```
max_error_t = 3.79831504821777
max_error_dist = 14.3955173492432
max_error_data = { "A": [22.0049724578857, -7.20710611343384, -3, 4.50648736953735, 0.53945195674896, 0.01614552922547], "B": [-0.80917793512344, -0.04843658953905], "a": 2.00355315208435, "b": 5.20355272293091, "loc_min_dist": 1.24090766906738, "loc_min_t": 3.79414415359497 }
max_error_result = (0, 15.63643)
```

The above output corresponds to the following [plot](https://sagecell.sagemath.org/?z=eJxVkMtuwkAMRfdI_IOVTUIbIs_YnkclFulvRBEKBCokSmiIumjVf6-nVSlsRvbxnetH300drOATsjp7gsbaCpGjtyw-BPElLH1l0Rt0xhATBVZEJXAl6Dh4clHIk5SAlVBkMVGc5xBdImicYREbrRX2bQnZc-qyxCpgNN5HEmOJk6eKOTA5CWoYUZK4U3EaiETIiEWta6Nso1h0KsXWWxtJvRQfh-369XBa94fLpApTWcaI3rmIzlO4UaQyVT6yYTZCEnVj-JrPar1Erwdp8jpv57P3bizyKV_MZ3st1A228AgTPEBRN-YnXuhTN7ZVdnkbJ-V0o-G_uG6kXahNd_Xvkv_mmm5Sej4OU7EvoSthk4zPw-E02b5ofjXX4XM9zT1KG-et4u1wHMZV_jLudqe8hMvhY7cSvDPDEoxUjhzT_4dx19_IvwF53IO-&lang=sage&interacts=eJyLjgUAARUAuQ==):

![to_be_minimal_or_not_to_be](https://user-images.githubusercontent.com/229837/225998591-fcd155e9-9fc5-4fdc-8e86-269d2d576e1e.png)

This shows the correct minimum in green (calculated using this PR), and the output of your minimization code in red. The red point is in fact on the graph of the function, but it is not in the interval where the minimum is sought, and it is also not a (local) minimum. I haven't yet examined your code in detail, but I suggest we first find out why it is not respecting the interval, and after that we can re-run the test project to iron out any remaining issues.

rburing committed:
Add Geometry3D::get_closest_points_between_circle_and_circle

Based on David Vranek's Fast and Accurate Circle-Circle and Circle-Line
3D Distance Computation.
Implementation based on the one in MinuteTorus by Sang Hyun Son, but
significantly modified.

rburing committed:
Add Geometry3D::get_closest_points_between_line_and_circle

Based on David Vranek's Fast and Accurate Circle-Circle and Circle-Line
3D Distance Computation.
Implementation based on the one once included in ODE by Russell L.
Smith, but heavily modified.

rburing commented:
I've updated/extended the copyright notices/licenses/credits for the circle-circle and line-circle closest points code.

I haven't touched the numerical minimization code yet (see the issues mentioned above).

Gamepro5 commented:
Looking forward to this; I noticed that the current implementation for cylinder collisions struggle both on trimesh collision shapes and the edges of a lot of shapes. For instance, if you are moving onto a different angled slope (hits the edge), it may misreport the  collision normal. If a pointy spike shape's vertex pokes the rounded part of the cylinder, it can get a wrong collision normal reported and essentially become stuck. Would like to help find edge cases with this implementation by doing extensive testing with it in my character controller.

_[*](https://github.com/Gamepro5/Custom_CharacterBody3D_Movement_System) I found that testing it in a character controller environment is quite effective, as any misrepresented move_and_collide() normal will immediately be obvious as the character will not move as expected, and will chain off into a noticeable error. It obviously isn't as credible as a unit test, but I've found a lot of collision detection bugs this way._

Gamepro5 commented:
So I did some testing. It seems like all this improves is the collision detection along the non-circular-cap parts of the cylinder.
[Here](https://youtu.be/AKTaLkE_dyU)'s a video of cylinder v cylinder collisions before on 4.0.2 stable, and [here](https://youtu.be/ef-SdqgKmP8)'s a video of cylinder v cylinder collisions on this build.

I have a feeling that the errors still present on the edges of the cylinder mesh might be the same issue as #70508
